modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C

```diff
@@ -0,0 +1,242 @@
+{
+  if (_flux_limiter_type == FluxLimiterTypeEnum::None)
+    return 0.0;
+  const Real p = PQPlusMinus(node_i, PQPlusMinusEnum::PPlus);
+  if (p == 0.0)
+    // Comment after Eqn (49): if P=0 then there's no antidiffusion, so no need to remove it
+    return 1.0;
+  const Real q = PQPlusMinus(node_i, PQPlusMinusEnum::QPlus);
+  const Real r = q / p;
+  return limitFlux(1.0, r);
+}
+
+Real
+AdvectiveFluxCalculator::rMinus(const Node & node_i) const
+{
+  if (_flux_limiter_type == FluxLimiterTypeEnum::None)
+    return 0.0;
+  const Real p = PQPlusMinus(node_i, PQPlusMinusEnum::PMinus);
+  if (p == 0.0)
+    // Comment after Eqn (49): if P=0 then there's no antidiffusion, so no need to remove it
+    return 1.0;
+  const Real q = PQPlusMinus(node_i, PQPlusMinusEnum::QMinus);
+  const Real r = q / p;
+  return limitFlux(1.0, r);
+}
+
+Real
+AdvectiveFluxCalculator::limitFlux(Real a, Real b) const
+{
+  if (_flux_limiter_type == FluxLimiterTypeEnum::None)
+    return 0.0;
+
+  if ((a >= 0.0 && b <= 0.0) || (a <= 0.0 && b >= 0.0))
+    return 0.0;
+  const Real s = (a > 0.0 ? 1.0 : -1.0);
+
+  const Real lal = std::abs(a);
+  const Real lbl = std::abs(b);
+  switch (_flux_limiter_type)
+  {
+    case FluxLimiterTypeEnum::MinMod:
+      return s * std::min(lal, lbl);
+    case FluxLimiterTypeEnum::VanLeer:
+      return s * 2 * lal * lbl / (lal + lbl);
+    case FluxLimiterTypeEnum::MC:
+      return s * std::min(0.5 * std::abs(a + b), 2.0 * std::min(lal, lbl));
+    case FluxLimiterTypeEnum::superbee:
+      return s * std::max(std::min(2.0 * lal, lbl), std::min(lal, 2.0 * lbl));
+    default:
+      return 0.0;
+  }
+}
+
+void
+AdvectiveFluxCalculator::pre()
+{
+  // Following algorithm sets _kij to zero for all nodes we can see
+
+  _kij = {};
+  // QUERY:: does the following loop run through nodes on this processor only (+ ghosted ones?) or
+  // all nodes?  We only want nodes on this processor
+  for (const auto & node_i : _mesh.getMesh().node_ptr_range())
+  {
+    std::map<Node, Real> nodal_flux = {};
+    std::vector<const Node *> neighbors;
+    MeshTools::find_nodal_neighbors(_mesh, *node_i, _nodes_to_elem_map, neighbors);
+    for (const auto & n : neighbors)
+      nodal_flux.emplace(*n, 0.0);
+    nodal_flux.emplace(*node_i, 0.0); // include node_i - node_i connection
+    _kij.emplace(*node_i, nodal_flux);
+  }
+}
+
+void
+AdvectiveFluxCalculator::computeElement()
+{
+  /// prepare variable values and test functions
+  _fe_problem.prepare(_current_elem, _tid);
+  _fe_problem.reinitElem(_current_elem, _tid);
+
+  /// compute _kij contributions from this element
+  for (unsigned i = 0; i < _current_elem->n_nodes(); ++i)
+  {
+    const Node & node_i = _current_elem->node_ref(i);
+    for (unsigned j = 0; j < _current_elem->n_nodes(); ++j)
+    {
+      const Node & node_j = _current_elem->node_ref(j);
+      for (unsigned qp = 0; qp < _qrule->n_points(); ++qp)
+        // KT Eqn (20)
+        _kij[node_i][node_j] +=
```

Comment:
Why not `node_i->id()`? You are using the whole `Node` class as an index + doing a memory copy here.

Reply:
Thanks for all your comments, @andrsd .    For this one (use `node_i->id()`), are you suggesting my _kij map should be
```
std::map<dof_id_type, std::map<dof_id_type, Real>>
```
instead of `std::map<Node, std::map<Node, Real>>` ?

Reply:
I wonder if `std::map<std::pair<dof_id_type, dof_id_type>, Real>` would be even better here.

Reply:
Decided against the std::pair idea.  I now use _kij to figure out the connected nodes, so the `std::map<dof_id_type, std::map<dof_id_type, Real>>` is better.  But i'm going to change _valence to use the std::pair.